### DISCUSSION TITLE
How to test React components that depend on usePage?

### DISCUSSION THREAD

**Discussion by pascaldevink** [inertiajs/inertia#675] [Help] [answered]
I have a single component that is using `usePage` that I want to test individually (using jest and react testing-library):
```
export default function Foobar() {
    const { authenticatedUser } = usePage().props

    return (
      <div>name: {authenticatedUser.name}</div>
    )
}
```

Normally, I would render the component in a test function, but `usePage` requires a `PageContext` to exist which is not exposed by the Inertia library...

I've tried wrapping the component in an `App`:
```
const pageProps = {
        props: {
            authenticatedUser: {
                name: "foo bar"
            }
        },
        url: "\/test",
        version: null
    }
    render(
        <App initialPage={pageProps} resolveComponent={() => {}}>
            <Foobar />
        </App>
    )

    // ... do some assertions
```

But that doesn't render anything:
```
<body>
  <div />
</body>
```

How would you test such a component that depends on an Inertia page to exist?

**Comment by pascaldevink** [comment]
@claudiodekker any tips you have on testing components that use `usePage` would be helpful 🙏

**Reply by claudiodekker** [reply]
@pascaldevink Hmm, React's not really my expertise, but from what I'm seeing you're using it correctly. Are you getting any errors/warnings at all? I'll see if I can reproduce this and provide a better answer later today.

**Reply by reinink** [reply]
Pinging @sebastiandedeyne, our React expert. 👍

**Comment by sebastiandedeyne** [comment]
Looking at your original snippet, this might work:

```js
const pageProps = {
    props: {
       authenticatedUser: {
           name: "foo bar"
       }
    },
    url: "\/test",
    version: null
}
    
render(
    <App
        initialPage={pageProps}
        resolveComponent={() => <Foobar />}
    />
)

// ... do some assertions
```

It's not ideal though, currently looking into how we can improve this :+1:

**Comment by sebastiandedeyne** [comment]
Another possible solution is to wrap `usePage` in your own hooks, and mock them in tests.

```js
function useAuthenticatedUser() {
  const { authenticatedUser } = usePage().props

  return authenticatedUser;
}

export default function Foobar() {
    const authenticatedUser = useAuthenticatedUser()

    return (
      <div>name: {authenticatedUser.name}</div>
    )
}

// In tests…
jest.mock('../useAuthenticatedUser', () => ({ name: 'Pascal' }))
```

**Reply by deleugpn** [reply]
I'm currently using this approach, but since `jest.mock` (`vi.mock`) has hoisting effect, this end up being quite a nightmare to manage among a large number of tests

**Comment by rhysemmerson** [comment, marked as answer]
You don't need to render the App component in order to test Foobar. Something like this will work

```js

jest.mock('@inertiajs/inertia-react', () => ({
  usePage: () => ({
    props: {
      authenticatedUser: {/* */}
    }
  })
}))

render(<Foobar />);

```

**Reply by pascaldevink** [reply]
This is exactly the route I took and this worked beautifully, thanks so much! 🙏

**Reply by progdog-ru** [reply]
Thank You! It helps me too

**Comment by choznerol** [comment]
For those who encountered the same error with [Storybook](https://storybook.js.org/), I found [mocking imports](https://storybook.js.org/docs/react/writing-stories/build-pages-with-storybook#mocking-imports) can be a solution.

<details>
<summary>Here is my config</summary>

```js
// .storybook/__mocks__/inertia-react.js
import * as InertiaReact from '@inertiajs/inertia-react';

let inertiaSharedProps;

export function inertiaReactDecorator(story, { parameters }) {
  if (parameters && parameters.inertia) {
    inertiaSharedProps = parameters.inertia.sharedProps;
  }
  return story();
}

const mockUsePage = () => {
  if (inertiaSharedProps) {
    return {
      props: inertiaSharedProps,
    };
  }
  inertiaSharedProps = null;
}

InertiaReact.usePage = mockUsePage;

export default InertiaReact
```
```js
// .storybook/main.js

module.exports = {
  ...
  webpackFinal: async (config) => {
    config.resolve.alias['inertia-react'] = require.resolve('./__mocks__/inertia-react.js');
    return config;
  },
};
```
```js
// .storybook/preview.js
import { inertiaReactDecorator } from './__mocks__/inertia-react';

export const decorators = [ ..., inertiaReactDecorator ]
```
```js
// Foo.stories.js
FooWithBarStatus.parameters = { inertia: { sharedProps: { isPublicPage: false } } };
```

</details>

P.s. I also tried wrapping the component with a `<InertiaApp>` at first but couldn't get it working 😅.
